docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,222 @@
+
+2. **Dispatch**: do work in response to the event
+
+3. **Leave**: perform any tasks that should be carried out after event handling (for example, tearing down any state set up during event enter)
+
+Before beginning to process an event, the event manager first determines which control should be the **target** of the event (taking into account visibilities, and bounding rectangles) [2].
+
+A list of "interested controls" is built.  This recursively contains every parent of the target.  This is used to determine which objects should have their event handlers run.
+
+**Notes:**
+
+1. The enter phase and leave phase facilitate the implementation of behaviour that wraps normal event dispatch.  For example, consider a composed widget with a border that highlights when the mouse is inside it.  To ensure that highlight changes sequence correctly with with its child widgets, the highlight must be enabled during the `MouseEnter` event's enter phase, and hidden during the `MouseLeave` event's leave phase.
+
+2. How the target of an event is determined may vary a lot on a per-event-type basis (for example, the target for a mouse click will depend on the mouse position, but the target for a keyboard event may depend on the keyboard focus).  For the purposes of this document, assume that it is an arbitrary process.
+
+### Event data
+
+When an event is submitted to the event manager, it has a **event name** and a set of **event data**.  The event data is an array containing all of the captured data relating to the event itself [1].
+
+In an event handler, `the event` evaluates to the event data array [2].
+
+Evaluating `the event` outwith an event handler is an error.
+
+`the event` is immutable.
+
+The `the event target` expression evaluates to the event's target control [3].
+
+**Notes:**
+
+1. For example, when creating a mouse event, it is important to capture the position of the mouse at the time the event began to be dispatched into the event data.  If computationally-intensive event handlers are run, the user could move the mouse between event creation and event handling.
+
+2. This permits simple implementation of syntactic sugar for frequently-needed event data.  For example:
+
+       the mouse position of the event
+
+   can easily be implemented as:
+
+       the event["mouse position"]
+
+
+3. It might be useful to provide `the event target is me` (or equivalent syntax) to test whether an event is targeted at the handling control or one of its children.
+
+
+### Event handlers
+
+All event handlers (for enter, dispatch and leave phases) must conform to:
+
+    public type EventHandler() returns any
+
+The return value of all event handlers is discarded.
+
+### Enter phase
+
+During the enter phase, **pre-event** handlers are run.  They are run from "outside-in"; a parent widget's pre-event handler runs before the pre-event handler of its child widgets.
```

Comment:
Okay, having written that there is some pondering to do here because it means that the widget event handling mental model does not quite subsume the current engine event handling model (in this case at least - the behavior of engine messages in this regard is very inconsistent).

Obviously for leaf widgets (those which have no children) pre/post are trivial/pointless - as you can do the pre/post actions in the on handler. Thus *perhaps* an alternative formulation would to change the meaning of 'pass':

```
on mouseUp
  do pre event stuff
  pass
  do post event stuff
end mouseUp
```

This is perhaps 'closer' to how the engine works for certain controls - but would mean that anything which 'contains' controls in LiveCode Script which wants to handle a contained control event *after* the contained control has had a go at it would have to do:

```
on mouseUp
  pass
  do post event stuff
end mouseUp
```

Perhaps something to ponder.